### DISCUSSION TITLE
Anonymous login seems not persistent

### DISCUSSION THREAD

**Discussion by Freecraft72** [GodotNuts/GodotFirebase#228] [Q&A] [answered]
I just setup my first small firebase project with your SDK and it works great :). I use it in the context of an asynchronous turn based game on Android. 
I'd like to allow a player to compete with other without necessary having to create an account (no login/pwd, google, or anything).
My understanding is that firebase anonymous login could do this, because from what I've read the UID generated seems to be tied to the device, and remain the same while the app isn't uninstalled from the phone. 
However this is not what I get for the moment : each time my app use "Firebase.Auth.login_anonymous()", I see that it got a new UUID in the firebase console. 
I have been looking for a parameter in the firebase console to control this but couldn't find anything. 

Did I misunderstand how anonymous login work ? 
Or maybe it is the way you implemented the SDK that generate this specific behavior ? 
Finally do you have any idea how I could/should implement what I want to do ? 

Thanks

**Comment by BearDooks** [comment]
@Freecraft72 Sorry I just saw this. Anonymous login should be tied to the UID of the device. I have seen it when I am testing Godot on my machine that it generates a new user every time I try. I believe once the app is exported it should stop doing this. @WolfgangSenff would know more as they actually used this in one of their apps. 

Small side note, if you are interested we started a discord for people to ask questions, show off work, and generally meet some other cool people using Godot. If you are interested. Invite to Discord: https://discord.gg/xXKYN2ZxDB

Thanks

Chuck

**Comment by WolfgangSenff** [comment, marked as answer]
Okay, so, apparently, the only reason anonymous login persists on the normal/supported Firebase SDKs is because they have implemented it that way. We have not - we have no code in there to support it in part because we can't always be guaranteed user storage with all exports, but also because...I thought it persisted too, but was wrong. :P We can open up an issue to investigate this and see what it would take.

**Reply by Freecraft72** [reply]
Thanks for the time you spent in checking this. It would be great if you could investigate this further ! 
Sadly my understanding is that there is not any workaround in order to get what I want. 
I hope you'll be able to implement this, otherwise I will not be able to use your SDK :( I will check your progress in resolving this issue. 
Thanks !

**Reply by WolfgangSenff** [reply]
I mean, you could just have them login properly and not anonymously. :) This doesn't really hinder you at all.

**Reply by Freecraft72** [reply]
I understand your point of view but login/registering can as well be too much effort for some people who just want to have a glance to a game. I know it's often the case for me for instance 😊

**Reply by WolfgangSenff** [reply]
Oh I get it - just saying, it doesn't prevent you from using the plugin. You can support both pretty easily, and just have anonymous be understood that it's not cached so they can't store their progress. Or you can implement a method of handling it server-side with a cloud function. There's a lot of ways to handle this that it just doesn't seem like you've considered is all.

**Reply by rool80** [reply]
I also ran into this problem. Every time I call the `Firebase.Auth.login_anonymous()` function, I always get a different UID. This means that this functionality in the plugin is completely useless, because the authors apparently didn’t understand what this function is actually supposed to do in real use.

When someone installs a game and the first thing that pops up is “register now,” they usually won’t even start playing. That’s why an anonymous account should be created, where the player’s progress is saved, and the next time the game is launched, it continues where it left off — because `Firebase.Auth.login_anonymous()` should log in using the same token/UID and the data can be loaded from Firebase.

If the player plays the game for, say, 5–10 days and then decides to create an account, these accounts are then merged and the player starts logging in with their own account. But you can’t force a player to log in with their own account right away!

So I repeat: in its current state, the `Firebase.Auth.login_anonymous()` function in this plugin is useless, and I have to keep looking further :(

**Reply by WolfgangSenff** [reply]
You're wrong and insulting, of course, and that's not appreciated - next strike and you're banned. You just have to call the save function and reload the save file when your game is ready for it. I as a library designer decided I couldn't make those decisions on someone's behalf because it's different with games vs business applications, so get over yourself.

https://github.com/GodotNuts/GodotFirebase/wiki/Authentication-and-User-Management#save-encrypted-auth-file


You can use this function to determine if you need to call to  login or if the auth file exists so you can load it: 

https://github.com/GodotNuts/GodotFirebase/blob/d960fecae1e45dbb192a04707e3ff83a84f3a5f4/addons/godot-firebase/auth/auth.gd#L394


If you do need to load the auth file, you then just call:

https://github.com/GodotNuts/GodotFirebase/blob/d960fecae1e45dbb192a04707e3ff83a84f3a5f4/addons/godot-firebase/auth/auth.gd#L493

**Reply by rool80** [reply]
You’re right, and I apologize to you. Thank you for the explanation and the example—that’s exactly what I needed, thanks.

**Comment by Freecraft72** [comment]
In case anyone would be in the same situation, I thought again to this and decided to implement what I want in simple (and ugly?) way. 
At first launch I automatically generate a unique_id/password, store this pair to a local file, and signin this user with email/password method.
While this local file exists my app will login using this id/pwd, otherwise it  create a new id/pwd , signin the user etc etc
It is probably awfull from a security point of view(maybe I could encrypt id/pwd ?) but for a small hobby project like mine it's ok.
